consensus/ibft/poa.go

```diff
@@ -0,0 +1,232 @@
+		if !params.snap.Set.Includes(params.header.Miner) {
+			return nil
+		}
+	}
+
+	voteCount := params.snap.Count(func(v *Vote) bool {
+		return v.Validator == params.proposer && v.Address == params.header.Miner
+	})
+
+	if voteCount > 1 {
+		// there can only be one vote per validator per address
+		return fmt.Errorf("more than one proposal per validator per address found")
+	}
+	if voteCount == 0 {
+		// cast the new vote since there is no one yet
+		params.snap.Votes = append(params.snap.Votes, &Vote{
+			Validator: params.proposer,
+			Address:   params.header.Miner,
+			Authorize: authorize,
+		})
+	}
+
+	// check the tally for the proposed validator
+	tally := params.snap.Count(func(v *Vote) bool {
+		return v.Address == params.header.Miner
+	})
+
+	// If more than a half of all validators voted
+	if tally > params.snap.Set.Len()/2 {
+		if authorize {
+			// add the candidate to the validators list
+			params.snap.Set.Add(params.header.Miner)
+		} else {
+			// remove the candidate from the validators list
+			params.snap.Set.Del(params.header.Miner)
+
+			// remove any votes casted by the removed validator
+			params.snap.RemoveVotes(func(v *Vote) bool {
+				return v.Validator == params.header.Miner
+			})
+		}
+
+		// remove all the votes that promoted this validator
+		params.snap.RemoveVotes(func(v *Vote) bool {
+			return v.Address == params.header.Miner
+		})
+	}
+
+	return nil
+}
+
+// candidateVoteHookParams are the params passed into the candidateVoteHook
+type candidateVoteHookParams struct {
+	header *types.Header
+	snap   *Snapshot
+}
+
+// candidateVoteHook checks if any candidate is up for voting by the operator
+// and casts a vote in the Nonce field of the block being built
+func (poa *PoAMechanism) candidateVoteHook(hookParams interface{}) error {
+	// Cast the params to candidateVoteHookParams
+	params := hookParams.(*candidateVoteHookParams)
+
+	// try to pick a candidate
+	if candidate := poa.ibft.operator.getNextCandidate(params.snap); candidate != nil {
+		params.header.Miner = types.StringToAddress(candidate.Address)
+		if candidate.Auth {
+			params.header.Nonce = nonceAuthVote
+		} else {
+			params.header.Nonce = nonceDropVote
+		}
+	}
+
+	return nil
+}
+
+// initializeHookMap registers the hooks that the PoA mechanism
+// should have
+func (poa *PoAMechanism) initializeHookMap() {
+	// Create the hook map
+	poa.hookMap = make(map[string]func(interface{}) error)
+
+	// Register the AcceptStateLogHook
+	poa.hookMap[AcceptStateLogHook] = poa.acceptStateLogHook
+
+	// Register the VerifyHeadersHook
+	poa.hookMap[VerifyHeadersHook] = poa.verifyHeadersHook
+
+	// Register the ProcessHeadersHook
+	poa.hookMap[ProcessHeadersHook] = poa.processHeadersHook
+
+	// Register the CandidateVoteHook
+	poa.hookMap[CandidateVoteHook] = poa.candidateVoteHook
+}
+
+// ShouldWriteTransactions indicates if transactions should be written to a block
+func (poa *PoAMechanism) ShouldWriteTransactions(blockNumber uint64) bool {
```

Comment:
Exactly.

Hooks are supposed to only execute logic and report an error in execution, not return values.

I felt like this method was a good compromise to be included as an interface method, because we might want to have additional logic in the future for deciding on block writes and this allows us to be flexible 👍